src/khepri_pattern_tree.erl

```diff
@@ -0,0 +1,270 @@
+      IsEmpty :: boolean(),
+      Payload :: payload().
+%% @doc Checks whether the given pattern tree is empty.
+%%
+%% A pattern tree node is empty if it contains no payload and has no children.
+%%
+%% @see empty/0.
+
+is_empty(#pattern_node{payload = ?NO_PAYLOAD, child_nodes = ChildNodes})
+  when ChildNodes =:= #{} ->
+    true;
+is_empty(_PatternTree) ->
+    false.
+
+-spec update(PatternTree, PathPattern, UpdateFun) -> Ret when
+      PatternTree :: khepri_pattern_tree:tree(Payload),
+      PathPattern :: khepri_path:native_pattern(),
+      UpdateFun :: update_fun(Payload),
+      Ret :: khepri_pattern_tree:tree(Payload),
+      Payload :: payload().
+%% @doc Updates the node of the given pattern tree with the given update
+%% function.
+%%
+%% If any pattern components of `PathPattern' do not yet exist in the tree,
+%% this function will add any necessary branches to the tree using empty tree
+%% nodes.
+%%
+%% If the tree node for the given `PathPattern' does not yet exist, the
+%% `?NO_PAYLOAD' constant will be passed to the `UpdateFun'.
+%%
+%% @see empty/0.
+
+update(PatternTree, [], UpdateFun) ->
+    update_payload(PatternTree, UpdateFun);
+update(
+  #pattern_node{child_nodes = ChildNodes0} = PatternTree,
+  [Component | Rest],
+  UpdateFun) ->
+    PatternSubtree = case ChildNodes0 of
+                         #{Component := PatternNode} ->
+                             PatternNode;
+                         _ ->
+                             empty()
+                     end,
+    PatternSubtree1 = update(PatternSubtree, Rest, UpdateFun),
+    ChildNodes = maps:put(Component, PatternSubtree1, ChildNodes0),
+    PatternTree#pattern_node{child_nodes = ChildNodes}.
+
+-spec update_payload(PatternTreeNode, UpdateFun) -> Ret when
+      PatternTreeNode :: khepri_pattern_tree:tree_node(Payload),
+      UpdateFun :: update_fun(Payload),
+      Ret :: khepri_pattern_tree:tree_node(Payload),
+      Payload :: payload().
+%% @doc Updates the payload of the given node with the given update function.
+%%
+%% @private
+
+update_payload(#pattern_node{payload = Payload} = PatternTree, UpdateFun) ->
+    Payload1 = UpdateFun(Payload),
+    PatternTree#pattern_node{payload = Payload1}.
+
+-spec fold(PatternTree, Tree, Path, FoldFun, Acc) -> Ret when
+      PatternTree :: khepri_pattern_tree:tree(Payload),
+      Tree :: khepri_tree:tree(),
+      Path :: khepri_path:native_path(),
+      FoldFun :: fold_fun(Payload),
+      Acc :: fold_acc(),
+      Ret :: fold_acc(),
+      Payload :: payload().
+%% @doc Folds over the given pattern tree to find all patterns in the tree
+%% which match the given path.
+%%
+%% The `FoldFun' function takes a path pattern, the payload stored in the
+%% pattern tree at that tree node, and an accumulator as arguments. Only tree
+%% nodes which have payloads are passed to this function.
+%%
+%% @see fold_fun().
+%% @see fold_acc().
+
+fold(PatternTree, Tree, Path, FoldFun, Acc) ->
```

Comment:
Probably `khepri_tree:find_matching_nodes/5` should be renamed `fold/5`. What it does changed a lot since its creation and the name remained the same.

Reply:
Do you think it's worth renaming `khepri_tree:find_matching_nodes/5` now? I think `fold/5` would fit better for that too

Reply:
Ok! I will create a separate PR with the rename 👍